compiler/rustc_middle/src/ty/context.rs

```diff
@@ -2769,8 +2769,33 @@ pub trait InternIteratorElement<T, R>: Sized {
 
 impl<T, R> InternIteratorElement<T, R> for T {
     type Output = R;
-    fn intern_with<I: Iterator<Item = Self>, F: FnOnce(&[T]) -> R>(iter: I, f: F) -> Self::Output {
-        f(&iter.collect::<SmallVec<[_; 8]>>())
+    fn intern_with<I: Iterator<Item = Self>, F: FnOnce(&[T]) -> R>(
+        mut iter: I,
+        f: F,
+    ) -> Self::Output {
+        // This code is hot enough that it's worth specializing for the most
+        // common length lists, to avoid the overhead of `SmallVec` creation.
+        // Lengths 0, 1, and 2 typically account for ~95% of cases. We assume
+        // that if the upper and lower bounds from `size_hint` agree they are
+        // correct.
```

Comment:
Can you constrain this method to ExactSizeIterator then?

Reply:
This came up when this code was first added: https://github.com/rust-lang/rust/pull/64949#discussion_r329975050

The current code will panic if the size_hint is wrong -- if it's too big then `iter.next().unwrap()` will panic, if it's too small then `assert!(iter.next().is_none())` will panic. Perhaps I should clarify what "assume" means here, it sounds weaker than the protection we've actually got.

Reply:
I have updated the comments.